src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs

```diff
@@ -0,0 +1,895 @@
+        InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+        OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<InvocationExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact]
+        public void CoalesceAssignment_CheckedDynamic()
+        {
+
+            string source = @"
+class C
+{
+    static void M(dynamic d1, dynamic d2)
+    {
+        checked
+        {
+            /*<bind>*/d1 ??= d2/*</bind>*/;
+        }
+    }
+}
+";
+            string expectedOperationTree = @"
+ICoalesceAssignmentOperation(IsChecked: True) (OperationKind.CoalesceAssignment, Type: dynamic) (Syntax: 'd1 ??= d2')
+  Target: 
+    IParameterReferenceOperation: d1 (OperationKind.ParameterReference, Type: dynamic) (Syntax: 'd1')
+  WhenNull: 
+    IParameterReferenceOperation: d2 (OperationKind.ParameterReference, Type: dynamic) (Syntax: 'd2')
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<AssignmentExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact]
+        public void CoalesceAssignmentFlow_SimpleAssignment()
+        {
+            string source = @"
+class C
+{
+    static void M(object o1, object o2)
+    /*<bind>*/{
+        o1 ??= o2;
+    }/*</bind>*/
+}
+";
+            string expectedFlowGraph = @"
+Block[B0] - Entry
+    Statements (0)
+    Next (Regular) Block[B1]
+        Entering: {R1} {R2}
+
+.locals {R1}
+{
+    CaptureIds: [1]
+    .locals {R2}
+    {
+        CaptureIds: [0]
+        Block[B1] - Block
+            Predecessors: [B0]
+            Statements (1)
+                IFlowCaptureOperation: 0 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'o1')
+                  Value: 
+                    IParameterReferenceOperation: o1 (OperationKind.ParameterReference, Type: System.Object) (Syntax: 'o1')
+
+            Jump if True (Regular) to Block[B3]
+                IIsNullOperation (OperationKind.IsNull, Type: System.Boolean, IsImplicit) (Syntax: 'o1')
+                  Operand: 
+                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: System.Object, IsImplicit) (Syntax: 'o1')
+
+            Next (Regular) Block[B2]
+        Block[B2] - Block
+            Predecessors: [B1]
+            Statements (1)
+                IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'o1')
+                  Value: 
+                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: System.Object, IsImplicit) (Syntax: 'o1')
```

Comment:
Why is there a flow capture of a flow capture here?

Reply:
Flow capture id 0 is just for the IsNull check. Flow capture 1 is for the evaluation of the expression.

---
In reply to: [207938024](https://github.com/dotnet/roslyn/pull/29063#discussion_r207938024) [](ancestors = 207938024)

Reply:
@AlekseyTs for an opinion here. This is an optimization we _could_ make, but I don't think we have another scenario in the CFG where we capture a result, do a test on the capture, and then re-capture in the false case.

---
In reply to: [207965573](https://github.com/dotnet/roslyn/pull/29063#discussion_r207965573) [](ancestors = 207965573,207938024)

Reply:
> This is an optimization we could make, but I don't think we have another scenario in the CFG where we capture a result, do a test on the capture, and then re-capture in the false case.

Yes, we intentionally avoid assigning to the same capture more than once on the same code path in the graph.

---
In reply to: [209018400](https://github.com/dotnet/roslyn/pull/29063#discussion_r209018400) [](ancestors = 209018400,207965573,207938024)